serve/mlc_serve/model/sampler.py

```diff
@@ -0,0 +1,472 @@
+    def _is_safe_to_sample(prob_like):
+        return (
+            torch.sum(torch.isnan(prob_like) | torch.isinf(prob_like) | (prob_like < 0))
+            == 0
+        )
+
+    res_greedy, res_random = None, None
+    sampling_tensors = sampling_metadata.sampling_tensors
+
+    batch_size = logits.shape[0]
+    mask_greedy_t, mask_random_t = (
+        sampling_tensors.mask_greedy,
+        sampling_tensors.mask_random,
+    )
+
+    if sampling_metadata.has_greedy:
+        logits_greedy = logits[mask_greedy_t]
+        res_greedy = torch.argmax(logits_greedy, -1)
+
+    probs_random = None
+    if sampling_metadata.has_random:
+        logits_random = logits[mask_random_t]
+        probs_random = torch.softmax(logits_random, dim=-1)
+        if check_safety and not _is_safe_to_sample(probs_random):
+            return None
+        res_random = torch.multinomial(probs_random, 1, True)[:, 0]
+
+    # Prepare output
+    # Send results to CPU and convert them into numpy
+    res_greedy = list(res_greedy.cpu().numpy()) if res_greedy is not None else list()
+    res_random = list(res_random.cpu().numpy()) if res_random is not None else list()
+
+    # Recover original order in the batch
+    # TODO: Can we vectorize this?
+    next_tokens = []
+    for batch_idx in range(batch_size):
+        sampling_info = sampling_metadata.index_map[batch_idx]
+        sampling_idx = sampling_info[1]
+        if sampling_info[0] == SamplingType.RANDOM:
+            assert sampling_idx < len(res_random)
+            next_tokens.append(res_random[sampling_idx])
+        else:
+            assert sampling_idx < len(res_greedy)
+            next_tokens.append(res_greedy[sampling_idx])
+
+    logprob_infos: List[Optional[RawLogprobsInfo]] = [None] * batch_size
+    if sampling_metadata.has_logprob:
+        all_top_logprobs = [torch.tensor([], device=logits.device)]
+        all_top_tokens = [torch.tensor([], device=logits.device)]
+        # If everything is random sampling, save one extra softmax
+        if not sampling_metadata.has_greedy:
+            assert probs_random is not None
+            logprobs = torch.log(probs_random)
+        else:
+            logprobs = torch.log_softmax(logits, dim=-1)
+
+        for lp_idx in range(LOGPROB_TOP_K_MAX):
+            logprob_topk = lp_idx + 1
+            mask_t = sampling_metadata.sampling_tensors.mask_top_logprob[logprob_topk]
+            top_logprobs, top_tokens = torch.topk(
+                logprobs[mask_t], k=logprob_topk, dim=-1, largest=True, sorted=True
+            )
+            all_top_logprobs.append(top_logprobs)
+            all_top_tokens.append(top_tokens)
+
+        # recover original batch order
+        # TODO: Can we vectorize this?
+        for batch_idx in range(batch_size):
+            is_logprobs, logprob_topk, idx = sampling_metadata.logprob_index_map[
+                batch_idx
+            ]
+            next_token = next_tokens[batch_idx]
+            if is_logprobs:
```

Comment:
Rather than looping over batch and check if each sequence needs logprob, it is better to create indices of sequence that needs logprob and loop over them. This was also what I insisted in https://github.com/octoml/mlc-llm/pull/82.

Probably you can create such indices tensor in `SamplingMetadata` in this new impl?

Reply:
Done. It has redundant computation but vectorized further. Throughput is slightly improved about 5% from the previous PR. The performance numbers are updated above. I see small degradation in the other entries, but these might be measurement noise.